src/Features/Core/Portable/Diagnostics/DiagnosticData.cs

```diff
@@ -198,39 +204,96 @@ public override string ToString()
                 Severity,
                 Message,
                 ProjectId,
-                MappedFilePath ?? "",
-                MappedStartLine,
-                MappedStartColumn,
-                OriginalFilePath ?? "",
-                OriginalStartLine,
-                OriginalStartColumn);
+                DataLocation?.MappedFilePath ?? "",
+                DataLocation?.MappedStartLine,
+                DataLocation?.MappedStartColumn,
+                DataLocation?.OriginalFilePath ?? "",
+                DataLocation?.OriginalStartLine,
+                DataLocation?.OriginalStartColumn);
+        }
+
+        public static Task<IEnumerable<Diagnostic>> ToDiagnosticsAsync(Project project, IEnumerable<DiagnosticData> datas, CancellationToken cancellationToken)
+        {
+            return ToDiagnosticsAsync(project, datas, documentIdToTree: null, cancellationToken: cancellationToken);
+        }
+
+        public static async Task<IEnumerable<Diagnostic>> ToDiagnosticsAsync(
+            Project project, IEnumerable<DiagnosticData> datas, ImmutableDictionary<DocumentId, SyntaxTree> documentIdToTree, CancellationToken cancellationToken)
+        {
+            var result = new List<Diagnostic>();
+            foreach (var data in datas)
+            {
+                result.Add(await data.ToDiagnosticAsync(project, documentIdToTree, cancellationToken).ConfigureAwait(false));
+            }
+            return result;
         }
 
-        public Diagnostic ToDiagnostic(SyntaxTree tree)
+        public Task<Diagnostic> ToDiagnosticAsync(Project project, CancellationToken cancellationToken)
         {
-            var location = Location.None;
-            if (tree != null)
+            return ToDiagnosticAsync(project, documentIdToTree: null, cancellationToken: cancellationToken);
+        }
+
+        private async Task<Diagnostic> ToDiagnosticAsync(Project project, ImmutableDictionary<DocumentId, SyntaxTree> documentIdToTree, CancellationToken cancellationToken)
+        {
+            return Diagnostic.Create(this.Id, this.Category, this.Message, this.Severity, this.DefaultSeverity, this.IsEnabledByDefault, this.WarningLevel, this.Title, this.Description, this.HelpLink, 
```

Comment:
The formatting of this statement is hard to read. More lines with some indenting might help.